src/DataSets.jl

```diff
@@ -516,41 +517,150 @@ function Base.open(f::Function, as_type, dataset::DataSet)
     end
 end
 
-# For convenience, this non-scoped open() just returns the data handle as
-# opened. See check_scoped_open for a way to help users avoid errors when using
-# this (ie, if `identity` is not a valid argument to open() because resources
-# would be closed before it returns).
+# Option 1
 #
-# FIXME: Consider removing this. It should likely be replaced with `load()`, in
-# analogy to FileIO.jl's load operation:
-# * `load()` is "load the entire file into memory as such-and-such type"
-# * `open()` is "open this resource, and run some function while it's open"
-Base.open(as_type, conf::DataSet) = open(identity, as_type, conf)
+# Attempt to use finalizers and the "async trick" to recover the resource from
+# inside a do block.
+#
+# Problems:
+# * The async stack may contain a reference to the resource, so it may never
+#   get finalized.
+# * The returned resource must be mutable to attach a finalizer to it.
 
-"""
-    check_scoped_open(func, as_type)
+#=
+function Base.open(as_type, dataset::DataSet)
+    storage_config = dataset.storage
+    driver = _drivers[storage_config["driver"]]
+    ch = Channel()
+    @async try
+        driver(storage_config, dataset) do storage
+            open(as_type, storage) do x
+                put!(ch, x)
+                if needs_finalizer(x)
+                    ch2 = Channel(1)
+                    finalizer(x) do _
+                        put!(ch2, true)
+                    end
+                    # This is pretty delicate — it won't work if the compiler
+                    # retains a reference to `x` due to its existence in local
+                    # scope.
+                    #
+                    # Likely, it depends on
+                    # - inlining the current closure into open()
+                    # - Assuming that the implementation of open() itself
+                    #   doesn't keep a reference to x.
+                    # - Assuming that the compiler (or interpreter) doesn't
+                    #   keep any stray references elsewhere.
+                    #
+                    # In all, it seems like this can't be reliable in general.
```

Comment:
@StefanKarpinski this is the technical reason that attempting to use finalizers to turn the existing `do`-based API into a finalizer based one won't work. This could certainly be hacked around by a breaking change to the DataSets storage driver API and converting some structs to be mutable in order to attach finalizers.

But more generally, there's a need to have "a place to hold references to resources" which can't be solved by finalizers and are solved by passing a resource context.